### ISSUE TITLE
DeepSpeed FP16 causes bert pretrain ACC decline or NAN? 

### ISSUE THREAD

**Issue by mpjlu** [microsoft/DeepSpeed#1267]
We use DeepSpeed (v0.4.0) FP16 to train Bert Large, we found some Acc decline cases or NAN cases.
Like this picture:
![image](https://user-images.githubusercontent.com/13826327/127647894-879f9239-838f-43ea-8170-852f19f73d6e.png)
And this picture:
![image](https://user-images.githubusercontent.com/13826327/127648300-0f493beb-9359-4ce0-b8b4-b258a4b512db.png)
When we set FP16 config to false or use AMP. The Acc convergence is good. 
DeepSpeed FP16 implementation is like apex O2, there is some convergence issue about amp O2.
Is there some solution to the convergence problem of DeepSpeed FP16?

**Comment by tjruwase** [comment]
@mpjlu, thanks for reporting this issue. Can you please share your deepspeed config?

**Comment by mpjlu** [comment]
My deepspeed config is:
{
  2   "gradient_accumulation_steps": 8,
  3   "train_micro_batch_size_per_gpu": 32,
  4   "steps_per_print": 1000,
  5   "prescale_gradients": false,
  6   "optimizer": {
  7     "type": "Lamb",
  8     "params": {
  9       "lr": 1e-3,
 10       "weight_decay": 0.01,
 11       "bias_correction": false,
 12       "max_coeff": 0.3,
 13       "min_coeff": 0.01
 14     }
 15   },
 16   "gradient_clipping": 1.0,
 17 
 18   "wall_clock_breakdown": false,
 19 
 20   "fp16": {
 21     "enabled": true,
 22     "loss_scale": 0
 23   }
 24 }
~       
Thanks

**Comment by tjruwase** [comment]
@mpjlu, thanks for sharing your deepspeed config. It looks reasonable. I have a few other questions to help understand the problem. 

1. What are those two plots showing? What are the axes? 
2. Can you share the corresponding amp convergence plot as well?
3. How many GPUs are you using?
4. Is the dataset publicly available for us to try to repro?

Thanks!

**Comment by tjruwase** [comment]
@mpjlu, just wanted to check if this was still an issue? Thanks.

**Comment by mpjlu** [comment]
Yes, this is still an issue. 
The plots show the steps (X) and the ACC,
The FP32 plot is like this:
![image](https://user-images.githubusercontent.com/13826327/130381234-59dab1b0-56df-44e1-ac4e-c20ecaf78ef0.png)

We are using 64 GPUs.
The dataset is not publicly available. 
DeepSpeed FP16 is like AMP O2, there are many discussions about the convergence issue of O2,  like this:
![image](https://user-images.githubusercontent.com/13826327/130381481-5ac5b798-29d2-46f3-8623-962e390de9ae.png)
https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html
So I want to know whether DeepSpeed FP16 has the same problem as amp O2.

**Comment by tjruwase** [comment]
@mpjlu, thanks for the clarifications. Yes, DeepSpeed FP16 is similar to amp O2 and should have same issues.

**Comment by mpjlu** [comment]
Do you have a plan to fix this problem?

**Comment by tjruwase** [comment]
@mpjlu, yes we would like to fix this problem. The challenge is that we don't currently have an easy repro. If you could provide such even on a small case, that would be very helpful.
